api/api_request_logger.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2024 The VeChainThor developers
+
+// Distributed under the GNU Lesser General Public License v3.0 software license, see the accompanying
+// file LICENSE or <https://www.gnu.org/licenses/lgpl-3.0.html>
+
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+
+	log "github.com/inconshreveable/log15"
+)
+
+// RequestLogger logs requests to an output
+type RequestLogger struct {
+	enabled      bool
+	writerChan   chan entry
+	stopChan     chan bool
+	outputWriter func(msg string, ctx ...interface{})
+}
+
+// NewThorRequestLogger applies thor business logic
+// when returning an instance of *RequestLogger
+func NewThorRequestLogger(enabledFlag bool, enableLog bool) *RequestLogger {
+	if enabledFlag {
+		return NewRequestLogger(enabledFlag, log.Info)
+	}
+	if enableLog {
+		return NewRequestLogger(enableLog, log.Debug)
+	}
+	return NewRequestLogger(false, nil)
+}
+
+func NewRequestLogger(enabled bool, logger func(msg string, ctx ...interface{})) *RequestLogger {
+	return &RequestLogger{
+		enabled:      enabled,
+		outputWriter: logger,
+		writerChan:   make(chan entry, 100_000),
+		stopChan:     make(chan bool),
+	}
+}
+
+func (l *RequestLogger) Enabled() bool {
+	return l.enabled
+}
+
+// Handle returns an http handler to ensure requests are syphoned into the writer
+func (l *RequestLogger) Handle(handler http.Handler) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		// Read and log the body (note: this can only be done once)
+		// Ensure you don't disrupt the request body for handlers that need to read it
+		var bodyBytes []byte
+		var err error
+		if r.Body != nil {
+			bodyBytes, err = io.ReadAll(r.Body)
+			if err != nil {
+				log.Warn("unexpected body read error", "err", err)
+				return // don't pass bad request to the next handler
+			}
+			r.Body = io.NopCloser(io.Reader(bytes.NewReader(bodyBytes)))
+		}
+
+		l.writerChan <- entry{
+			Timestamp: time.Now(),
+			URI:       r.URL.String(),
+			Method:    r.Method,
+			Body:      string(bodyBytes),
+		}
+
+		// call the original http.Handler we're wrapping
+		handler.ServeHTTP(w, r)
+	}
+
+	// start the writer
+	l.start()
+
+	// http.HandlerFunc wraps a function so that it
+	// implements http.Handler interface
+	return http.HandlerFunc(fn)
+}
+
+func (l *RequestLogger) Stop() {
+	if !l.enabled {
+		return
+	}
+
+	// make sure any pending message is written
+	l.stopChan <- true
+	<-l.stopChan
+}
+
+func (l *RequestLogger) start() {
+	go func() {
+		for {
+			select {
+			case item := <-l.writerChan:
```

Comment:
Instead of building a queue inside the request logger, how about delegating the queue to the logger function, I found that logger handler can be configured to a buffered handler.

Reply:
I don't mind, but would like to avoid marshalling overhead on the request proxy, would that still be possible ?